sklearn/tree/_utils.pyx

```diff
@@ -301,3 +303,328 @@ cdef class PriorityHeap:
 
         return 0
+
+# =============================================================================
+# WeightedPQueue data structure
+# =============================================================================
+
+cdef class WeightedPQueue:
+    """A priority queue class, always sorted in increasing order.
+
+    Attributes
+    ----------
+    capacity : SIZE_t
+        The capacity of the array
+
+    array_ptr : SIZE_t
+        The water mark of the array; the array grows from left to right in the
+        array ``array_``. array_ptr is always less than capacity.
+
+    array_ : WeightedPQueueRecord*
+        The array of array records. The minimum element is on the left;
+        the array grows from left to right
+    """
+
+    def __cinit__(self, SIZE_t capacity):
+        self.capacity = capacity
+        self.array_ptr = 0
+        self.array_ = <WeightedPQueueRecord*> calloc(capacity, sizeof(WeightedPQueueRecord))
+
+        if self.array_ == NULL:
+            raise MemoryError()
+
+    def __dealloc__(self):
+        free(self.array_)
+
+    cdef bint is_empty(self) nogil:
+        return self.array_ptr <= 0
+
+    cdef SIZE_t size(self) nogil:
+        return self.array_ptr
+
+    cdef int push(self, DOUBLE_t data, DOUBLE_t weight) nogil:
+        """Push record on the array.
+        Returns 0 if successful; -1 on out of memory error.
+        """
+        cdef SIZE_t array_ptr = self.array_ptr
+        cdef WeightedPQueueRecord* array = NULL
+        cdef SIZE_t i
+
+        # Resize if capacity not sufficient
+        if array_ptr >= self.capacity:
+            self.capacity *= 2
+            array = <WeightedPQueueRecord*> realloc(self.array_,
+                                                    self.capacity *
+                                                    sizeof(WeightedPQueueRecord))
+
+            if array == NULL:
+                # no free; __dealloc__ handles that
+                return -1
+            self.array_ = array
+
+        # Put element as last element of array
+        array = self.array_
+        array[array_ptr].data = data
+        array[array_ptr].weight = weight
+
+        # bubble last element up according until it is sorted
+        # in ascending order
+        i = array_ptr
+        while(i != 0 and array[i].data < array[i-1].data):
+            array[i], array[i-1] = array[i-1], array[i]
+            i -= 1
+
+        # Increase element count
+        self.array_ptr = array_ptr + 1
+        return 0
+
+    cdef int remove(self, DOUBLE_t value, DOUBLE_t weight) nogil:
+        """Remove a specific value/weight record from the array.
+        Returns 0 if successful, -1 if record not found."""
+        cdef SIZE_t array_ptr = self.array_ptr
+        cdef WeightedPQueueRecord* array = self.array_
+        cdef SIZE_t idx_to_remove = -1
+        cdef SIZE_t i
+
+        if array_ptr <= 0:
+            return -1
+
+        # find element to remove
+        for i in range(array_ptr):
```

Comment:
so for the reference / others, I implemented binary search here and it turned out to be 3x slower than sequential search. We're sticking with sequential for now.